File: Kanban-Board/user_service/src/main/java/com/niit/user_service/service/UserService.java
package com.niit.user_service.service;

import com.niit.user_service.domain.Board;
import com.niit.user_service.domain.User;
import com.niit.user_service.exception.UserAlreadyExistException;
import com.niit.user_service.exception.UserNotFoundException;

import java.util.List;

public interface UserService {
    User saveUser(User user) throws UserAlreadyExistException;

    User findById(String userId) throws UserNotFoundException;

    User getUserByEmail(String emailId) throws UserNotFoundException;

    void updateUser(User user) throws UserNotFoundException;

    void deleteUser(String userId) throws UserNotFoundException;

    List<User> findAll();

    User findByEmailId(String emailId) throws UserNotFoundException;
    List<Board> getBoardsByEmail(String emailId) throws UserNotFoundException;

    void updateUserProfilePic(String userId, String profilePicId) throws UserNotFoundException;

    boolean checkUserExistsByEmail(String email);
}

File: Kanban-Board-Frontend/src/app/services/user.service.ts
// user.service.ts

import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { User } from '../interface/User';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = 'http://localhost:11111/user';

  constructor(private http: HttpClient) { }

  // Add user
  public addUser(user: any) {
    return this.http.post<any>(`${this.apiUrl}/save`, user);
  }

  getUserById(userId: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/${userId}`);
  }

  // Fetch user by emailId
  getUserByEmail(emailId: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/email/${emailId}`);
  }

  // Fetch all users
  getAllUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}`);
  }

  // Update user
  updateUser(userId: string, user: User): Observable<any> {
    return this.http.put(`${this.apiUrl}/${userId}`, user);
  }

  // Delete user
  deleteUser(userId: string): Observable<any> {
    return this.http.delete(`${this.apiUrl}/${userId}`);
  }

  updateUserProfilePic(userId: string, formData: FormData): Observable<any> {
    return this.http.put(`${this.apiUrl}/${userId}/profilePic`, formData);
  }
 
}
